### PR TITLE
feat(shared): Introduce client-side caching for telemetry events

### PR DESCRIPTION
**Pull request by LauraBeatris**
## Description

Originally raised (https://github.com/clerk/javascript/pull/3267/files) but had to re-open here due GitHub permissions issues on running integrations from a forked branch.

Currently blocking SDK-1672 (https://github.com/clerk/javascript/pull/3257) and https://github.com/clerk/javascript/pull/3279

Introduces client-side caching for telemetry events. This prevents event flooding in frequently executed code paths, such as for React hooks or components.

Event sampling by rate will be executed on the server-side, or when neither the browser storage or window are available. 

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/shared/src/telemetry/clientCache.ts
Comment (LauraBeatris) [inline packages/shared/src/telemetry/clientCache.ts:29]: Realized this logic is incorrect. When an entry expires, we should update `clerk_telemetry` cache object and remove that key from it. 

When returning here, an event is cached when it isn't expired & it's entry exists.

File: packages/shared/src/telemetry/clientCache.ts
Comment (BRKalow) [inline packages/shared/src/telemetry/clientCache.ts:5]: Is this value just for testing? Or is the comment wrong? 👀

File: packages/shared/src/telemetry/clientCache.ts
Comment (BRKalow) [inline packages/shared/src/telemetry/clientCache.ts:11]: nit: Do you think like `TelemetryEventThrottler`, or something similar, would be a better name? I feel like that more clearly communicates the intent 💭

File: packages/shared/src/telemetry/clientCache.ts
Comment (BRKalow) [inline packages/shared/src/telemetry/clientCache.ts:15]: if we change the naming this could be: `isEventThrottled`

File: packages/shared/src/telemetry/collector.ts
Comment (BRKalow) [inline packages/shared/src/telemetry/collector.ts:117]: We might want to pass `preparedPayload` here so we can include it as part of the throttled key, as if any of the "global" metadata that is provided in that method changes the event would still get throttled.

File: packages/shared/src/__tests__/telemetry.test.ts
Comment (BRKalow) [inline packages/shared/src/__tests__/telemetry.test.ts:215]: Nice tests 🥇

**Review by BRKalow — COMMENTED** [review]
Looking good! Left a comment about the naming of the cache class. I wonder if calling it a "throttler" (or something) is more accurate, as opposed to a cache.

File: packages/shared/src/telemetry/collector.ts
Comment (BRKalow) [inline packages/shared/src/telemetry/collector.ts:137]: Can we include the `isStorageSupported` check in `cacheAndRetrieve()`? That would simplify usage here

File: packages/shared/src/telemetry/clientCache.ts
Comment (BRKalow) [inline packages/shared/src/telemetry/clientCache.ts:72]: ❓ why did we do property access with brackets instead of `window.localStorage`?

File: packages/shared/src/telemetry/clientCache.ts
Comment (LauraBeatris) [reply packages/shared/src/telemetry/clientCache.ts:5]: Ooops... I was testing it and ended pushing the wrong value.

File: packages/shared/src/telemetry/clientCache.ts
Comment (LauraBeatris) [reply packages/shared/src/telemetry/clientCache.ts:11]: Yeah, it makes sense to me! `ClientCache` is indeed a bit generic, it leads to the question: "caching for what"?

File: packages/shared/src/telemetry/collector.ts
Comment (LauraBeatris) [reply packages/shared/src/telemetry/collector.ts:117]: Yeah, it makes sense! Better rely on the `preparedPayload` than the raw one.

File: .changeset/metal-bulldogs-destroy.md
Comment (LekoArts) [inline .changeset/metal-bulldogs-destroy.md:5]:
```suggestion
Internal change to add client-side caching to Clerk's telemetry. This prevents event flooding in frequently executed code paths, such as for React hooks or components. Gracefully falls back to old behavior if e.g. `localStorage` is not available. As such, no public API changes and user changes required.
```

You can follow [these guidelines](https://www.notion.so/clerkdev/Code-collaboration-4d32380fb78c4226abd9e1bcb7d55abf?pvs=4#62d500ed4eea46c0b8732f8fe4db6056) to write more thorough changesets. I'm also not always following them 1:1 but they are good general guardrails.

File: packages/shared/src/telemetry/clientCache.ts
Comment (panteliselef) [reply packages/shared/src/telemetry/clientCache.ts:15]: Yeap this name feels more appropriate, my understanding here is that if a key needs to be invalidated we just delete and expect the next attempt to cache it again

File: packages/shared/src/__tests__/telemetry.test.ts
Comment (LekoArts) [inline packages/shared/src/__tests__/telemetry.test.ts:171]: Nit: You could also move the `fetchSpy` init to a `let` variable and create the mock in `beforeEach` like with `window.get`

**Review by panteliselef — APPROVED** [review]
This looks good to me, I'm approving in order not to block this PR further, suggestions from @BRKalow probably should be addressed before merging. I've added a few questions myself. 

Amazing stuff

**Review by LekoArts — COMMENTED** [review]
Great work on this!

File: packages/shared/src/telemetry/clientCache.ts
Comment (LekoArts) [reply packages/shared/src/telemetry/clientCache.ts:45]: By sorting it we can ensure that payloads with the same content (but with different ordering) are still cached in the same key

Comment (LauraBeatris) [reply packages/shared/src/telemetry/clientCache.ts:45]: Will add a JSDocs on this method to clarify about that!

File: packages/shared/src/__tests__/telemetry.test.ts
Comment (LauraBeatris) [reply packages/shared/src/__tests__/telemetry.test.ts]: I've moved some spies to the root `describe` block!

File: packages/shared/src/telemetry/throttler.ts
Comment (LauraBeatris) [inline packages/shared/src/telemetry/throttler.ts:48]: Just need to refactor this according to @BRKalow feedback, we should use the `preparedPayload` rather than the raw one.

**Review by BRKalow — APPROVED** [review]
⭐

**Pull request merged**
